docs/CODE_GUIDELINES.md

```diff
@@ -0,0 +1,174 @@
+    boundary. It might be a whole microservice, or a chunk of a monolith that
+    is internally consistent.
+- User-facing messages (and log messages):
+  - Put the variable part of a message at the end, preceded by a colon.
+    - `'foo failed: {0}'`
+  - If a message has many variable parts, mention them in the message in
+    a natural way, but surround them with quotes:
+    - `'creating "{0}" in directory: {1}'`
+- Bubble-up error conditions, do not sink them to random hidden places (such as
+  logs only), expecting callers to figure out the failure mode. If a caller
+  spawns a process that fails, the caller should get either an exception or
+  a callback or a return value indicating that some descendant in the call-chain
+  failed, with the failure reason.
+- Refactoring tools allow us to avoid "premature abstraction". Avoid wrappers
+  until the need is clear and obvious.
+  - Example: What is the purpose of `CloudFormation.Resource`, what is the
+    purpose vs SDK types?
+
+## Test guidelines
+
+- Test real codepaths, avoid mocks.
+  - Example 1: instead of mocking various aspects of the `sam local invoke`
+    codepath, assert the string representation of the CLI command generated by
+    `samCliBuild.ts`
+  - Example 2: instead of mocking various aspects of `sam local invoke`
+    codepath, shift all CONFIG decisions earlier in the codepath and test the
+    real values.
+  - Both examples result in faster, more-readable, more-useful tests, and more
+    test-coverage of real codepaths.
+
+## Code guidelines
+
+- Use maps to group variables, rather than ad-hoc/informal group signals. [Example](https://github.com/aws/aws-toolkit-vscode/blob/abed2c4c7e1329da785190e286e567525afa9da5/src/test/shared/utilities/timeoutUtils.test.ts#L107-L113)
+  - PREFER:
+    ```
+    let topic = {
+      var1 = 'foo',
+      var2 = 'bar',
+    }
+    ```
+  - INSTEAD OF:
+    ```
+    let topicVar1 = 'foo'
+    let topicVar2 = 'bar'
+    ```
+- Where formal grouping is not possible, using a common prefix in names helps
+  discoverability and helps other developers understand "these symbols are
+  related".
+  - PREFER:
+    ```
+    export interface AwsSamDebuggerConfigLoose extends AwsSamDebuggerConfig {
+      ...
+    }
+    ```
+  - INSTEAD OF:
+    ```
+    export interface MorePermissiveAwsSamDebuggerConfig extends AwsSamDebuggerConfig {
+      ...
+    }
+    ```
+- Use module-qualified names, to avoid aliasing members.
+  - PREFER:
+    ```
+    import * as foo from '../foo'
+    const result = foo.Result
+    ```
+  - INSTEAD OF:
+    ```
+    import { Result as FooResult } from '../foo'
+    const result = FooResult
+    ```
+- Use small names for small scopes.
+  - PREFER:
+    ```
+    things.filter(o => o.isFoo)
+    ```
+  - INSTEAD OF:
+    ```
+    things.filter(thing => thing.isFoo)
+    ```
+
+## User settings
+
+- Global scope is shared across all vscode instances *including* remote/SSH
```

Comment:
Unsure if you're confusing Global State and contributed settings? Global State (accessible by `vscode.ExtensionContext.globalState`) is not accessible to the user: https://code.visualstudio.com/api/extension-capabilities/common-capabilities#data-storage , whereas contributed settings are (those are accessible via `vscode.workspace.getConfiguration`): https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema

We do have some degree of infrasturcture around contributed settings (see `DefaultSettingsConfiguration`) but we don't around global state values; propose some guidelines for naming global state values?

Reply:
global (settings) scope != global state